From 2ce99bdc9fb5b92170c1e51ca71eae8130ecb66f Mon Sep 17 00:00:00 2001
From: Ma Wupeng <mawupeng1@huawei.com>
Date: Fri, 22 Apr 2022 15:46:56 +0800
Subject: [PATCH 3543/4585] [Huawei] mm: Introduce reliable_debug=S to control
 shmem use mirrored memory

Offering: HULK
hulk inclusion
category: feature
bugzilla: 187270

------------------------------------------

Introduce reliable_debug=S to control shmem use mirrored memory.

Signed-off-by: Ma Wupeng <mawupeng1@huawei.com>
---
 Documentation/admin-guide/kernel-parameters.txt | 3 ++-
 mm/mem_reliable.c                               | 4 ++++
 2 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/Documentation/admin-guide/kernel-parameters.txt b/Documentation/admin-guide/kernel-parameters.txt
index 9342c945845a..db5ee2c0056c 100644
--- a/Documentation/admin-guide/kernel-parameters.txt
+++ b/Documentation/admin-guide/kernel-parameters.txt
@@ -4781,11 +4781,12 @@
 			See Documentation/admin-guide/cgroup-v1/cpusets.rst.
 
 	reliable_debug=	[ARM64]
-			Format: [F][,P]
+			Format: [F][,S][,P]
 			Only works with CONFIG_MEMORY_RELIABLE and
 			"kernelcore=reliable" is configured.
 			F: User memory allocation(special user task, tmpfs) will
 			not allocate memory from non-mirrored region if failed.
+			S: The shmem does not use the reliable memory.
 			P: Page cache does not use the reliable memory.
 
 	reserve=	[KNL,BUGS] Force kernel to ignore I/O ports or memory
diff --git a/mm/mem_reliable.c b/mm/mem_reliable.c
index b1788e27290e..2ed60d8b100d 100644
--- a/mm/mem_reliable.c
+++ b/mm/mem_reliable.c
@@ -351,6 +351,10 @@ static int __init setup_reliable_debug(char *str)
 			reliable_allow_fallback = false;
 			pr_info("disable memory reliable fallback\n");
 			break;
+		case 'S':
+			shmem_reliable = false;
+			pr_info("disable shmem use reliable memory\n");
+			break;
 		case 'P':
 			pagecache_use_reliable_mem = false;
 			pr_info("disable page cache use reliable memory\n");
-- 
2.17.1

